### PR TITLE
Add logging mechanism

### PR DESCRIPTION
**Pull request by lpichler**
This adds support to unified logging. This adds possibility to control all loggers with same settings, formatting, output etc.

This adds:
- **logrus** logger with setting, formatting
  -  possibility to set log level:
    -  key in config :`LogLevel` (or `LOG_LEVEL` as env. var.)
    - usage:

 ```go
import (
	logging "github.com/RedHatInsights/sources-api-go/logger"
)

logging.Log.Debug("test")
logging.Log.Warn("test")
logging.Log.Info("test") //=>
/*
{
"@timestamp":"2021-08-04T16:16:48.551Z",
"@version":1,
"app":
"source-api-go",
"caller":"main.main",
"filename":"/Users/lpichler/projects/topological/sources-api-go/main.go:19",
"hostname":"lpichler-mac",
"labels":{"app":"source-api-go"},
"level":"info",
"message":"TEST",
"tags":["source-api-go"]
}
*/
```

- Support logrus in **echo** - this available in `echo.context` (for example in handlers or routes)
  - usage:
```go
// ./source_handlers.go:29
func SourceList(c echo.Context) error {
	c.Logger().Info("tenant: 1") //=>
/*
{"@timestamp":"2021-08-04T16:25:21.305Z",
"@version":1,"app":
"source-api-go",
"hostname":"lpichler-mac",
"labels":{"app":"source-api-go"},
"level":"info",
"message":"tenant: 1",
"tags":["source-api-go"]}
*/
``` 
 - NOTE: there is no `filename` and `caller`

- support for logging echo's middleware - logging requests - updated only formatting
  -  possibility to set log level for middleware's log messages: (if `LogLevelForMiddlewareLogs` is `DEBUG` so middleware will display log messages only when `LogLevel` will be set to `DEBUG` )
    -  key in config :`LogLevelForMiddlewareLogs`
  - example:
```go
/*
{"@timestamp":"2021-08-04T16:16:48.56Z",
"@version":1,
"app":"source-api-go",
"bytes_in":"19",
"bytes_out":"998",
"error":"",
"host":"localhost:8000",
"hostname":"lpichler-mac",
"id":"",
"labels":{"app":"source-api-go"},
"latency":"1929487",
"latency_human":"1.929487ms",
"level":"DEBUG",
"method":"GET",
"remote_ip":"::1",
"status":"200",
"tags":["source-api-go"],
"time":"2021-08-04T16:33:43.816303+02:00",
"uri":"/api/sources/v3.1/sources",
"user_agent":"PostmanRuntime/7.28.2"}
*/
```

- support **logrus** in GORM
  -  possibility to set log level for sql log messages: (if `LogLevelForSqlLogs` is `DEBUG` so GORM will display sql log messages only when `LogLevel` will be set to `DEBUG` )
    -  key in config :`LogLevelForSqlLogs`
  - possibility to set threshold for slow queries
    - key in config :`SlowSQLThreshold ` and value integer (seconds)
    - example: when `SlowSQLThreshold` is 2 then queries which take more than 2 second are logged
  - example:
```go
/*
{"@timestamp":"2021-08-04T16:33:43.816Z",
"@version":1,
"app":"source-api-go",
"duration":0.579647,
"filename":"/Users/lpichler/projects/topological/sources-api-go/dao/source_dao.go:30",
"hostname":"lpichler-mac",
"labels":{"app":"source-api-go"},
"level":"info",
"message":"SELECT * FROM \"sources\" WHERE tenant_id = 1 AND \"sources\".\"paused_at\" IS NULL LIMIT 100",
"rows":3,
"tags":["source-api-go"]}
*/
```

--- 

Links
https://issues.redhat.com/browse/RHCLOUD-15004

### CODE REVIEW

**Comment by lpichler** [comment]
cc @syncrou

**Review by syncrou — COMMENTED** [review]
This was very interesting. I don't get a lot of it, but a small bit. Thanks for copying me on this @lpichler

File: dao/db.go [hunk 1]
```diff
@@ -2,17 +2,33 @@ package dao
...
 
 func Init() {
-	db, err := gorm.Open(postgres.Open(dbString()), &gorm.Config{})
+	logger := &logrus.Logger{
+		Out:       logging.LogOutputFrom(conf.LogHandler),
+		Level:     logging.LogrusLogLevelFrom(conf.LogLevel),
+		Formatter: logging.NewCustomLoggerFormatter(conf, true),
+	}
+
+	l := &logging.CustomGORMLogger{SkipErrorRecordNotFound: true,
```
Comment (lindgrenj6) [inline dao/db.go]: can we move the `SkipErrorRecordNotFound` field to the next line? I'm surprised the formatter allows that.

**Review by lindgrenj6 — COMMENTED** [review]
Just one tiny comment, otherwise looks good. Interesting that echo has its own logger interface, I initially thought it used the same one as the standard logging package.

**Comment by lpichler** [comment]
@lindgrenj6 fixed, thanks for review 👍

**Review by lindgrenj6 — APPROVED** [review]

**Pull request merged**